tls/s2n_client_hello.c

```diff
@@ -307,10 +307,14 @@ int s2n_client_hello_recv(struct s2n_connection *conn)
 
     /* Call client_hello_cb if exists, letting application to modify s2n_connection or swap s2n_config */
     if (conn->config->client_hello_cb) {
-        if (conn->config->client_hello_cb(conn, conn->config->client_hello_cb_ctx) < 0) {
+        int rc = conn->config->client_hello_cb(conn, conn->config->client_hello_cb_ctx);
+        if (rc < 0) {
             GUARD(s2n_queue_reader_handshake_failure_alert(conn));
             S2N_ERROR(S2N_ERR_CANCELLED);
         }
+        if (rc) {
```

Comment:
Can you instead just set this on the s2n_connection inside the client_hello_cb? I think there are use cases for the client_hello_cb that do not relate to SNI.

I think we also have to be careful about changing the default behavior of the server. There are servers out there that have clients that are advertising the SNI extension in the ClientHello and the server is currently acting on the extension and not sending the response in the ServerHello. If we change it to always start sending the response (instead of allowing this to be a configurable behavior), there are clients that may not correctly parse the server response and will no longer be able to successfully complete a handshake. It would be nice to have this new behavior be toggled by config.

Reply:
if the `client_hello_cb ` returns 0 then the behavior isn't changed.  per the previous API documentation, it should be returning 0 or a negative value to indicate error, which means, as it stands now, no current servers will change their behavior.

Reply:
Ignore the comment about default behavior since the client_hello_cb can just return 0. That's the second time today I've forgotten that < 0 and > 0 does not include the = 0 case =)